Mathlib/Topology/Compactification/OnePointEquiv.lean

```diff
@@ -0,0 +1,149 @@
+
+/-!
+# One-point compactification and projectivization
+
+We construct a set-theoretic equivalence between
+`OnePoint K` and the projectivization `ℙ K (Fin 2 → K)` for an arbitrary field `K`.
+
+(This equivalence can be extended to a homeomorphism
+in the case `K = ℝ`, where `OnePoint ℝ` gets the topology of one-point compactification.
+This result is to be added in a different file.)
+
+## Main definitions
+
+* `divOnePoint`: Division on `K` taking values in `OnePoint K`. Notation: `÷`.
+* `divSlope`: Lowering `divOnePoint` to a map `ℙ K (Fin 2 → K) → OnePoint K`.
+
+## Main results
+
+* `divSlope_equiv` proves that `ℙ K (Fin 2 → K) ≃ OnePoint K`.
+
+## Tags
+
+one-point extension, projectivization
+-/
+
+open scoped LinearAlgebra.Projectivization OnePoint
+open Projectivization Classical
+
+/-!
+### Definition and basic properties
+
+We define a division operation `divOnePoint`, with notation `÷`.
+It takes inputs from a field `K` and outputs a value in `K ∪ {∞}`, i.e.,  `OnePoint K`.
+We prove that `÷` preserves projective equivalence: `(c•a)÷(c•b)=a÷b`.
+Hence `÷` projects down to a map `divSlope` from `ℙ K (Fin 2 → K)` to `OnePoint K`
+that takes a line through the origin and returns its slope. For a vertical line
+the slope is `∞`.
+-/
+variable {K : Type*}
+/-- A modified division from a structure with `Zero` and `Div` to its
+ `OnePoint` extension. With notation `÷` for `divOnePoint`,
+  we have in particular that `1 ÷ 0 = ∞`. -/
+noncomputable def divOnePoint [Zero K] [Div K] (a : K) (r : K): OnePoint K :=
+  if r ≠ 0 then a / r else ∞
+
+/-- Uncurried version of `divOnePoint`, with nonzeroness assumption. -/
+noncomputable def divOnePoint' [Zero K] [Div K]
+  (u : {v : Fin 2 → K // v ≠ 0}) : OnePoint K :=
+  if u.1 1 ≠ 0 then some (u.1 0 / u.1 1) else ∞
+
+/-- Notation `÷` showing that `divOnePoint` is distinct from
+ ordinary division `/`. -/
+infix:50 " ÷ " => divOnePoint
+
+/-- `divOnePoint` can be lifted to the projective line (see `divSlope`.) -/
+lemma divOnePoint_lifts [Field K] (a b : {v : Fin 2 → K // v ≠ 0})
+    (h : ∃ c : Kˣ, (fun m : Kˣ ↦ m.1 • b.1) c = a.1) :
+    (a.1 0 ÷ a.1 1) = (b.1 0 ÷ b.1 1) := by
+  obtain ⟨c,hc⟩ := h
+  rw [← hc]
+  simp only [divOnePoint, ne_eq, Fin.isValue, Pi.smul_apply, ite_not]
+  split_ifs with hbc hb hb
+  · rfl
+  · simp only [ne_eq, OnePoint.infty_ne_coe]
+    exact hb <| (Units.mul_right_eq_zero c).mp hbc
+  · rw [hb] at hbc
+    simp at hbc
+  · apply congrArg some
+    field_simp
+    ring
+
+/-- Equivalence between the projective line and the one-point extension. -/
+noncomputable def divSlope [Field K] (p : ℙ K (Fin 2 → K)) : OnePoint K :=
+  Quotient.lift (fun u => divOnePoint (u.1 0) (u.1 1)) divOnePoint_lifts p
+
+/-! ### Equivalence
+We establish the equivalence between `OnePoint K` and `ℙ K (Fin 2 → K)` for a field `K`.
+-/
+
+/-- Extensionality for functions from `Fin 2`. -/
+lemma funext_fin2 {a b : Fin 2 → K} (h₀ : a 0 = b 0) (h₁ : a 1 = b 1) : a = b :=
```

Comment:
This can surely go in a much more basic file (note that `Fin` is defined before mathlib, but just find a mathlbi file that works)?

Reply:
Hmm... is it okay to just incorporate the proof `funext fun j => by fin_cases j <;> simp[hc,ga]` where the lemma is used, like this?
```
      have : a.1 0 ≠ 0 := fun hc => .elim <|a.2 <|funext fun j => by fin_cases j <;> simp[hc,ga]
      have : b.1 0 ≠ 0 := fun hc => .elim <|b.2 <|funext fun j => by fin_cases j <;> simp[hc,gb]
```